main/src/sphexa/propagator.hpp

```diff
@@ -172,132 +179,160 @@ class HydroProp final : public Propagator<DomainType, ParticleDataType>
-        computeMomentumAndEnergy(first, last, ngmax_, d, domain.box());
-        timer.step("MomentumEnergyIAD");
+        timer.step("  + mpi::synchronizeHalos: c11, c12, c13, c22, c23, c33");
+#ifdef USE_CUDA
+        CHECK_CUDA_ERR(cudaMemcpy(d.devPtrs.d_c11, d.c11.data(), size_np_T, cudaMemcpyHostToDevice));
+        CHECK_CUDA_ERR(cudaMemcpy(d.devPtrs.d_c12, d.c12.data(), size_np_T, cudaMemcpyHostToDevice));
+        CHECK_CUDA_ERR(cudaMemcpy(d.devPtrs.d_c13, d.c13.data(), size_np_T, cudaMemcpyHostToDevice));
+        CHECK_CUDA_ERR(cudaMemcpy(d.devPtrs.d_c22, d.c22.data(), size_np_T, cudaMemcpyHostToDevice));
+        CHECK_CUDA_ERR(cudaMemcpy(d.devPtrs.d_c23, d.c23.data(), size_np_T, cudaMemcpyHostToDevice));
+        CHECK_CUDA_ERR(cudaMemcpy(d.devPtrs.d_c33, d.c33.data(), size_np_T, cudaMemcpyHostToDevice));
+        timer.step("  * GPU CudaCopyIn Sync HostToDevice: c11, c12, c13, c22, c23, c33");
+#endif
+
+        computeDivvCurlv(first, last, ngmax_, d, domain.box());
+#ifdef USE_CUDA
+        CHECK_CUDA_ERR(cudaMemcpy(d.divv.data(), d.devPtrs.d_divv, size_np_T, cudaMemcpyDeviceToHost));
+        CHECK_CUDA_ERR(cudaMemcpy(d.curlv.data(), d.devPtrs.d_curlv, size_np_T, cudaMemcpyDeviceToHost));
+#endif
+        timer.step("VelocityDivCurl");
+        domain.exchangeHalos(d.divv, d.curlv);
+        timer.step("  + mpi::synchronizeHalos: divv, curlv");
+#ifdef USE_CUDA
+        CHECK_CUDA_ERR(cudaMemcpy(d.devPtrs.d_divv, d.divv.data(), size_np_T, cudaMemcpyHostToDevice));
+        CHECK_CUDA_ERR(cudaMemcpy(d.devPtrs.d_curlv, d.curlv.data(), size_np_T, cudaMemcpyHostToDevice));
+        timer.step("  * GPU CudaCopyIn Sync HostToDevice: divv, curlv");
+#endif
+
+        computeAVswitches(first, last, ngmax_, d, domain.box());
+#ifdef USE_CUDA
+        CHECK_CUDA_ERR(cudaMemcpy(d.alpha.data(), d.devPtrs.d_alpha, size_np_T, cudaMemcpyDeviceToHost));
+#endif
+        timer.step("AVswitches");
+        domain.exchangeHalos(d.alpha);
+        timer.step("  + mpi::synchronizeHalos: alpha");
+#ifdef USE_CUDA
+        CHECK_CUDA_ERR(cudaMemcpy(d.devPtrs.d_alpha, d.alpha.data(), size_np_T, cudaMemcpyHostToDevice));
+        timer.step("  * GPU CudaCopyIn Sync HostToDevice: alpha");
+#endif
+
+        computeMomentumEnergy(first, last, ngmax_, d, domain.box());
+#ifdef USE_CUDA
+        CHECK_CUDA_ERR(cudaMemcpy(d.du.data(), d.devPtrs.d_du, size_np_T, cudaMemcpyDeviceToHost));
+
+        if (d.g == 0)
+        {
+            CHECK_CUDA_ERR(cudaMemcpy(d.grad_P_x.data(), d.devPtrs.d_grad_P_x, size_np_T, cudaMemcpyDeviceToHost));
+            CHECK_CUDA_ERR(cudaMemcpy(d.grad_P_y.data(), d.devPtrs.d_grad_P_y, size_np_T, cudaMemcpyDeviceToHost));
+            CHECK_CUDA_ERR(cudaMemcpy(d.grad_P_z.data(), d.devPtrs.d_grad_P_z, size_np_T, cudaMemcpyDeviceToHost));
+        }
+#endif
+        timer.step("MomentumEnergy");
 
         d.egrav = 0.0;
         if (doGrav)
         {
             mHolder_.upsweep(d, domain);
```

Comment:
We should add here a comment stating that gravity modifies gradP adding the gravitational forces.